### PR TITLE
ODSP Test Driver: Fallback to browser auth when local

### PR DESCRIPTION
**Pull request by anthony-murphy**

### CODE REVIEW

**Review by jatgarg — APPROVED** [review]

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 1]
```diff
@@ -65,13 +74,23 @@ export class OdspTestDriver implements ITestDriver {
         const emailServer = username.substr(username.indexOf("@") + 1);
         const server = `${emailServer.substr(0, emailServer.indexOf("."))}.sharepoint.com`;
 
+        const directoryParts = [config?.directory];
+
+        // if we are in a azure dev ops build use the build id in the dir path
+        if (process.env.BUILD_BUILD_ID !== undefined) {
+            directoryParts.push(process.env.BUILD_BUILD_ID);
+        } else {
+            directoryParts.push(os.hostname());
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:83]: why add this to the path? Is this machine name or something?

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 2]
```diff
@@ -65,13 +74,23 @@ export class OdspTestDriver implements ITestDriver {
...
+
         return this.create(
             {
                 username,
                 password: passwords[username],
                 server,
+                inAutomation: process.env.BUILD_BUILD_ID !== undefined,
             },
-            config?.directory ?? "",
+            directoryParts.filter((p)=>p).join("/"),
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:93]: when is `p` undefined?

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 3]
```diff
@@ -116,14 +127,37 @@ export class OdspTestDriver implements ITestDriver {
...
+            // Where the test driver is for a specific site.
+            const tokens = await this.odspTokenManager.getOdspTokens(
+                new URL(options.siteUrl).hostname,
+                config,
+                passwordTokenConfig(config.username, config.password),
+                options.refresh,
+            );
+            return tokens.accessToken;
+        }catch(ex) {
+            if(config.inAutomation === true) {
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:142]: cool

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 4]
```diff
@@ -116,14 +127,37 @@ export class OdspTestDriver implements ITestDriver {
...
     ) {
-        // This function can handle token request for any multiple sites. Where the test driver is for a specific site.
-        const tokens = await this.odspTokenManager.getOdspTokens(
-            new URL(options.siteUrl).hostname,
+        const hostname = new URL(options.siteUrl).hostname;
+        try{
+            // This function can handle token request for any multiple sites.
+            // Where the test driver is for a specific site.
+            const tokens = await this.odspTokenManager.getOdspTokens(
+                new URL(options.siteUrl).hostname,
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:135]: use `hostname` var?

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 5]
```diff
@@ -116,14 +127,37 @@ export class OdspTestDriver implements ITestDriver {
...
+        const browserTokens = await this.odspTokenManager.getOdspTokens(
+            hostname,
             config,
-            passwordTokenConfig(config.username, config.password),
-            options.refresh,
+            {
+                type: "browserLogin",
+                navigator: (openUrl)=>{
+                    // eslint-disable-next-line max-len
+                    console.log(`Open the following url in a new private browser window, and login with user: ${config.username}`);
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:153]: Maybe add a hint to check that env variable for the password?

File: packages/test/test-drivers/src/odspTestDriver.ts [hunk 6]
```diff
@@ -116,14 +127,37 @@ export class OdspTestDriver implements ITestDriver {
...
+                    console.log(`"${openUrl}"`);
+                },
+            },
+            true,
+            true,
         );
-        return tokens.accessToken;
+        return browserTokens.accessToken;
```
Comment (markfields) [inline packages/test/test-drivers/src/odspTestDriver.ts:160]: Does this actually work?  Don't you need an auth redirect service running at localhost?

**Review by markfields — APPROVED** [review]
🎂 LGTM

**Pull request merged**